cmd/expiration-mailer/main.go

```diff
@@ -205,88 +212,122 @@ func (m *mailer) processCerts(ctx context.Context, allCerts []core.Certificate)
 		regIDToCerts[cert.RegistrationID] = cs
 	}
 
-	err := m.mailer.Connect()
-	if err != nil {
-		return fmt.Errorf("connecting to SMTP server: %w", err)
+	var wg sync.WaitGroup
+	workChan := make(chan work)
+
+	parallelSends := m.parallelSends
+	if parallelSends == 0 {
+		parallelSends = 1
 	}
-	defer func() {
-		_ = m.mailer.Close()
-	}()
 
+	for senderNum := uint(0); senderNum < parallelSends; senderNum++ {
+		// For politeness' sake, don't open more than 1 new connection per
+		// second.
+		if senderNum > 0 {
+			time.Sleep(time.Second)
+		}
+
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		conn, err := m.mailer.Connect()
+		if err != nil {
+			m.log.AuditErrf("connecting parallel sender %d: %s", senderNum, err)
+			close(workChan)
+			return err
+		}
+		wg.Add(1)
+		go func(conn bmail.Conn, ch <-chan work) {
+			defer wg.Done()
+			for w := range ch {
+				err := m.sendToOneRegID(ctx, conn, w.regID, w.certs)
+				if err != nil {
+					m.log.AuditErr(err.Error())
+				}
+			}
+			conn.Close()
+		}(conn, workChan)
+	}
 	for regID, certs := range regIDToCerts {
+		workChan <- work{regID, certs}
```

Comment:
We don't start sending work to the `workChan` until after we're done spinning up all of the connections and worker goroutines. We sleep for a whole second each time we go through the loop above. So if we have `parallelSends = 100`, it'll take us 100 seconds before we even start writing work to the channel for those 100 goroutines to consume.

Reply:
Good point. I reworked this so populating the work channel happens in a goroutine that starts _before_ the sender goroutines. To ensure that goroutine doesn't block forever (if the sender goroutines all error out), I increased the size of workChan to fit all the certificates. This is a bit of an unfortunate hack, since that's a bit of a big size, but we already buffer the whole set of certificates in memory a few different times in this code, so it's not too bad.